tools/state-viewer/src/latest_witnesses.rs

```diff
@@ -57,3 +82,53 @@ impl LatestWitnessesCmd {
         }
     }
 }
+
+#[derive(Parser)]
+pub struct ValidateWitnessCmd {
+    /// File with state witness saved as vector in JSON.
+    #[arg(long)]
+    input_file: PathBuf,
+}
+
+impl ValidateWitnessCmd {
+    pub(crate) fn run(&self, home_dir: &Path, near_config: NearConfig, store: Store) {
+        let encoded_witness: Vec<u8> =
+            serde_json::from_reader(BufReader::new(std::fs::File::open(&self.input_file).unwrap()))
+                .unwrap();
+        let witness: ChunkStateWitness = borsh::BorshDeserialize::try_from_slice(&encoded_witness)
+            .expect("Failed to deserialize witness");
+        let chain_genesis = ChainGenesis::new(&near_config.genesis.config);
+        let epoch_manager =
+            EpochManager::new_arc_handle(store.clone(), &near_config.genesis.config);
+        let runtime_adapter =
+            NightshadeRuntime::from_config(home_dir, store, &near_config, epoch_manager.clone())
+                .expect("could not create the transaction runtime");
+        let shard_tracker = ShardTracker::new(
+            TrackedConfig::from_config(&near_config.client_config),
+            epoch_manager.clone(),
+        );
+        // TODO(stateless_validation): consider using `ChainStore` instead of
+        // `Chain`.
+        let chain = Chain::new_for_view_client(
+            Clock::real(),
+            epoch_manager.clone(),
+            shard_tracker,
+            runtime_adapter.clone(),
+            &chain_genesis,
+            DoomslugThresholdMode::TwoThirds,
+            false,
+        )
+        .unwrap();
+        chain
+            .shadow_validate_state_witness(
+                witness,
+                epoch_manager.as_ref(),
+                runtime_adapter.as_ref(),
+            )
+            .unwrap();
+        // Not optimal, but needed to wait for potential `validate_state_witness` failure.
+        loop {
+            std::thread::sleep(core::time::Duration::from_secs(60));
+        }
+    }
```

Comment:
:cry: 

Maybe you could add a `ProcessingDoneTracker` argument to `shadow_validate_state_witness`, like it's done in `process_chunk_state_witness`/`start_validating_chunk`

https://github.com/near/nearcore/blob/62bc785f2a94cae7ecd72ec61363bb57793c458b/chain/client/src/stateless_validation/chunk_validator/mod.rs#L115